File: examples/go/table/agg/handlers/start_hand.go
package handlers

import (
	"crypto/sha256"
	"encoding/binary"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardStartHand(state TableState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Table does not exist")
	}
	if state.Status == "in_hand" {
		return angzarr.NewCommandRejectedError("Hand already in progress")
	}
	if state.ActivePlayerCount() < 2 {
		return angzarr.NewCommandRejectedError("Not enough players to start hand")
	}
	return nil
}

func computeHandStarted(state TableState, tableRoot []byte) *examples.HandStarted {
	handNumber := state.HandCount + 1
	handRoot := generateHandRoot(tableRoot, handNumber)

	dealerPosition := advanceToNextActive(state.DealerPosition, state)
	smallBlindPosition := advanceToNextActive(dealerPosition, state)
	bigBlindPosition := advanceToNextActive(smallBlindPosition, state)

	var activePlayers []*examples.SeatSnapshot
	for _, seat := range state.Seats {
		if !seat.IsSittingOut {
			activePlayers = append(activePlayers, &examples.SeatSnapshot{
				Position:   seat.Position,
				PlayerRoot: seat.PlayerRoot,
				Stack:      seat.Stack,
			})
		}
	}

	return &examples.HandStarted{
		HandRoot:           handRoot,
		HandNumber:         handNumber,
		DealerPosition:     dealerPosition,
		SmallBlindPosition: smallBlindPosition,
		BigBlindPosition:   bigBlindPosition,
		ActivePlayers:      activePlayers,
		GameVariant:        state.GameVariant,
		SmallBlind:         state.SmallBlind,
		BigBlind:           state.BigBlind,
		StartedAt:          timestamppb.New(time.Now()),
	}
}

// HandleStartHand handles the StartHand command.
func HandleStartHand(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state TableState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.StartHand
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardStartHand(state); err != nil {
		return nil, err
	}

	event := computeHandStarted(state, commandBook.Cover.Root.Value)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}

// generateHandRoot creates a deterministic hand root from table root and hand number.
func generateHandRoot(tableRoot []byte, handNumber int64) []byte {
	h := sha256.New()
	h.Write(tableRoot)
	buf := make([]byte, 8)
	binary.BigEndian.PutUint64(buf, uint64(handNumber))
	h.Write(buf)
	return h.Sum(nil)
}

// advanceToNextActive finds the next active (non-sitting-out) player position.
func advanceToNextActive(currentPos int32, state TableState) int32 {
	maxPlayers := state.MaxPlayers
	for i := int32(1); i <= maxPlayers; i++ {
		nextPos := (currentPos + i) % maxPlayers
		if seat, exists := state.Seats[nextPos]; exists && !seat.IsSittingOut {
			return nextPos
		}
	}
	return currentPos // Shouldn't happen if we have active players
}

File: examples/rust/table/agg/src/handlers/start_hand.rs
//! StartHand command handler.

use sha2::{Sha256, Digest};

use angzarr_client::proto::examples::{HandStarted, SeatSnapshot, StartHand};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::TableState;

fn guard(state: &TableState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Table does not exist"));
    }
    if state.status == "in_hand" {
        return Err(CommandRejectedError::new("Hand already in progress"));
    }
    if state.active_player_count() < 2 {
        return Err(CommandRejectedError::new("Not enough players to start hand"));
    }
    Ok(())
}

fn compute(state: &TableState, table_root: &[u8]) -> HandStarted {
    let hand_number = state.hand_count + 1;
    let hand_root = generate_hand_root(table_root, hand_number);

    let dealer_position = advance_to_next_active(state.dealer_position, state);
    let small_blind_position = advance_to_next_active(dealer_position, state);
    let big_blind_position = advance_to_next_active(small_blind_position, state);

    let active_players: Vec<SeatSnapshot> = state
        .seats
        .values()
        .filter(|seat| !seat.is_sitting_out)
        .map(|seat| SeatSnapshot {
            position: seat.position,
            player_root: seat.player_root.clone(),
            stack: seat.stack,
        })
        .collect();

    HandStarted {
        hand_root,
        hand_number,
        dealer_position,
        small_blind_position,
        big_blind_position,
        active_players,
        game_variant: state.game_variant as i32,
        small_blind: state.small_blind,
        big_blind: state.big_blind,
        started_at: Some(angzarr_client::now()),
    }
}

pub fn handle_start_hand(
    command_book: &CommandBook,
    command_any: &Any,
    state: &TableState,
    seq: u32,
) -> CommandResult<EventBook> {
    let _cmd: StartHand = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;

    let table_root = command_book
        .cover
        .as_ref()
        .and_then(|c| c.root.as_ref())
        .map(|u| u.value.as_slice())
        .unwrap_or(&[]);

    let event = compute(state, table_root);
    let event_any = pack_event(&event, "examples.HandStarted");

    Ok(new_event_book(command_book, seq, event_any))
}

/// Generate deterministic hand root from table root and hand number.
fn generate_hand_root(table_root: &[u8], hand_number: i64) -> Vec<u8> {
    let mut hasher = Sha256::new();
    hasher.update(table_root);
    hasher.update(hand_number.to_be_bytes());
    hasher.finalize().to_vec()
}

/// Find the next active (non-sitting-out) player position.
fn advance_to_next_active(current_pos: i32, state: &TableState) -> i32 {
    let max_players = state.max_players;
    for i in 1..=max_players {
        let next_pos = (current_pos + i) % max_players;
        if let Some(seat) = state.seats.get(&next_pos) {
            if !seat.is_sitting_out {
                return next_pos;
            }
        }
    }
    current_pos // Shouldn't happen if we have active players
}
